File: core/src/com/oni/donjon/data/GameSave.java
package com.oni.donjon.data;

import com.badlogic.gdx.math.Vector2;
import com.oni.donjon.map.TileType;
import lombok.Data;

/**
 * @author Daniel Chesters (on 07/02/16).
 */
public class GameSave {
    private int mapHeight;
    private int mapWidth;
    private SavedTile[][] map;
    private Vector2 playerPosition;


    public GameSave() {
        this.playerPosition = Vector2.Zero;
    }

    public GameSave(int mapHeight, int mapWidth, SavedTile[][] map,
        Vector2 playerPosition) {
        this.mapHeight = mapHeight;
        this.mapWidth = mapWidth;
        this.map = map;
        this.playerPosition = playerPosition;
    }

    public Vector2 getPlayerPosition() {
        return playerPosition;
    }

    public void setPlayerPosition(Vector2 playerPosition) {
        this.playerPosition = playerPosition;
    }

    public SavedTile[][] getMap() {
        return map;
    }

    public void setMap(SavedTile[][] map) {
        this.map = map;
    }

    public int getMapHeight() {
        return mapHeight;
    }

    public void setMapHeight(int mapHeight) {
        this.mapHeight = mapHeight;
    }

    public int getMapWidth() {
        return mapWidth;
    }

    public void setMapWidth(int mapWidth) {
        this.mapWidth = mapWidth;
    }


    @Data
    public static class SavedTile {
        private TileType type;
        private boolean know;

        public SavedTile(TileType type, boolean know) {
            this.type = type;
            this.know = know;
        }

        public SavedTile() {
            //Empty for JSON processor
        }
    }
}

File: core/src/main/kotlin/com/oni/donjon/data/GameSave.kt
package com.oni.donjon.data

import com.badlogic.gdx.math.Vector2
import com.oni.donjon.map.TileType
import java.util.*

/**
 * @author Daniel Chesters (on 07/02/16).
 */
data class GameSave(
        var mapHeight: Int? = null,
        var mapWidth: Int? = null,
        var map: Array<Array<SavedTile?>>? = null,
        var playerPosition: Vector2 = Vector2.Zero) {
    data class SavedTile(
            var type: TileType? = null,
            var know: Boolean? = false)

    override operator fun equals(other: Any?): Boolean {
        when (other) {
            is GameSave -> {
                for (x in 0..mapWidth!!) {
                    for (y in 0..mapHeight!!) {
                        if (map?.get(x)?.get(y) != other.map?.get(x)?.get(y)) {
                            return false
                        }
                    }
                }

                return other.mapHeight == mapHeight
                        && other.mapWidth == mapWidth
                        && other.playerPosition == playerPosition
            }
            else -> return false
        }
    }

    override fun hashCode(): Int {

        return Objects.hash(mapHeight, mapWidth, map, playerPosition)
    }
}
